### PR TITLE
Raise toolchain requirements for MXNet 2 

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
As per https://github.com/apache/incubator-mxnet/issues/17968, require C++17 compatible compiler. 
For cuda code, use C++14 mode introduced in Cuda 9. In general there is no issue with linking together objects resulting from different standards as long as they are compiled by the same compiler. 

This PR contains the following specific changes:
- Switch CI pipeline to use gcc7 on Ubuntu and CentOS 
- Switch CD pipeline to CentOS 7 with https://www.softwarecollections.org/en/scls/rhscl/devtoolset-7/ This enables us to build with gcc7 C++17 compiler while keeping a relatively old glibc requirement for distribution.
- Simplify ARM Edge builds
  - Switch to standard Ubuntu / Debian cross-compilation toolchain for ARMv7, ARMv8
  - Switch to https://toolchains.bootlin.com/ toolchain for ARMv6 (the Debian ARMv6 toolchain is for ARMv4 + ARMv5 + ARMv6, but we wish to only target ARMv6 and make use of ARMv6 features)
  - Remove reliance on `dockcross` for cross compilation.
- Simplify Jetson build
  - Use standard Ubuntu / Debian cross-compilation toolchain for ARMv8
  - Upgrade to Cuda 10 and Jetpack 4.3
  - Simplify build setup
- Simplify QEMU ARM virtualization test setup on CI
  - Remove complex "Virtual Machine in Docker" logic and run a QEMU based Docker container instead based on `arm32v7/ubuntu`
- Fix out of bounds vector accesses in
  - SoftmaxGradOpType
  - MKLDNNFCBackward
- Fix use of non-standard `rand_r` function (which is not available on anymore on newer Android toolchains and shouldn't be use in any case).
- Fix reproducibility of RNN with Dropout
- Fix reproducibility of DGL Graph Sampling Operators
- Update tests for Android Edge build to NDK19. The [previously used standalone toolchain is obsolete](https://developer.android.com/ndk/guides/standalone_toolchain).

Those Dockerfiles that required refactoring as part of the effort were refactored based on the following consideration
- Maximize the use of system dependencies provided by the distribution instead of manually installing dependencies from source or from third party vendors. This reduces the complexity of the installation process and essentially pins the dependency versions, increasing CI stability. Further, Dockerfile build speed is improved. To facilitate this, use recent distribution versions. We still ensure backwards compatibility via CentOS7 based build and test stages
- Minimize the number of layers in the Dockerfile. Don't have 5 different script files executed, each calling `apt-get update` and `install`, but just execute once. Speeds up the build and reduces image size. Keep each Dockerfile simple and tailored to a purpose, instead of running 20 scripts to install dependencies for every thinkable scenario, which is unmaintainable.
 
Some more small changes:
- Remove outdated references to Cuda 7 and Cuda 8 in various files.
- Remove C++03 support in mshadow
- Disable broken tests (may fix as part of this PR if I find the time)
  - NumpyBooleanAssignForwardCPU https://github.com/apache/incubator-mxnet/issues/17990
  - test_init.test_rsp_const_init https://github.com/apache/incubator-mxnet/issues/17988

### CODE REVIEW

**Comment by ptrendx** [comment]
Note: With this we will be able to change shared_ptr here: https://github.com/apache/incubator-mxnet/blob/master/src/engine/threaded_engine.h#L440 to `unique_ptr` (as it was supposed to be but C++11 did not allow `std::move` to lambda).

**Comment by leezu** [comment]
@larroy this relies on the binfmt fix binary support of linux kernel: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/binfmt_misc.txt?h=linux-4.8.y#n69 There is also a small summary at https://codepyre.com/2019/12/arming-yourself/

**Comment by leezu** [comment]
@mxnet-bot run ci [unix-cpu]

Due to https://github.com/apache/incubator-mxnet/issues/17067

File: src/operator/contrib/dgl_graph.cc [hunk 1]
```diff
@@ -24,6 +24,9 @@
 #include <mxnet/operator_util.h>
```
Comment (szha) [inline src/operator/contrib/dgl_graph.cc:24]: @zheng-da could you help review this in #17984?

Comment (zheng-da) [reply src/operator/contrib/dgl_graph.cc:24]: DGL no longer depends on the contrib graph operators. I would suggest to deprecate this sampling operator and even remove it in MXNet 2.0. Please see my comment here. https://github.com/apache/incubator-mxnet/issues/16167#issuecomment-613160299

**Pull request merged**